protolog/buffer.go

```diff
@@ -0,0 +1,26 @@
+package protolog
+
+import (
+	"io"
+)
+
+type Buffer struct {
+	reader io.Reader
+	buffer []byte
+}
+
+func NewBuffer(reader io.Reader) *Buffer {
+	return &Buffer{reader, make([]byte, 0)}
+}
+
+func (buffer *Buffer) Read(p []byte) (n int, err error) {
+	n, err = buffer.reader.Read(p)
+	buffer.buffer = append(buffer.buffer, p[:n]...)
+	return n, err
+}
+
+func (buffer *Buffer) String() string {
+	data := buffer.buffer
+	buffer.buffer = make([]byte, 0)
```

Comment:
Rather than allocating new memory, reduce the slice to zero length so we can reuse the memory next time round:
`buffer.buffer = buffer.buffer[:0]`